Several more bug fixes to Tofino code generation

anirudhSK opened the pull request:
1. Fixes xors not being supported by Tofino (#20).
2. Fixes the fact that tofino.alu was not consistent with what was supported by the hardware. This is similar to the concerns raised in https://github.com/chipmunk-project/chipmunk-tofino/pull/16. (https://github.com/chipmunk-project/chipmunk-tofino/commit/8b1091f1deab85f3522476714c754f3c237b76ac and https://github.com/chipmunk-project/chipmunk-tofino/commit/2db38dbdb806e7dac2c138ac9370811b9244cc3c)
3. Ensures that all state variables and packet fields are signed integers (https://github.com/chipmunk-project/chipmunk-tofino/commit/8464b2ffd0ebcfa2175dd7ed369f146d099542cc and https://github.com/chipmunk-project/chipmunk-tofino/commit/da7385d36266d83922b1c55d29da2709184a0ee5).
4. Ensure that table names are short (https://github.com/chipmunk-project/chipmunk-tofino/commit/f2d5a716c85133e1374faec1344df1eb259a09ce and https://github.com/chipmunk-project/chipmunk-tofino/commit/f0b4a0320a0c6ddea1efe800620f846aff51e2f5)

taegyunkim commented on templates/stateless_alu.j2:
```diff
@@ -7,7 +7,9 @@ int {{ alu_name }}({{arg_list|join(',')}}, int opcode_hole_local, int immediate_
   if (opcode == 0) {
-    return (alu_mode == 0 ? x + y : (alu_mode == 1 ? x + immediate_operand : immediate_operand));
+    return (alu_mode == 0 ? x + y
+                        : (alu_mode == 1 ? x + immediate_operand
```
This is unnecessary, because x and y are coming from the same set of inputs.

taegyunkim commented on templates/stateless_alu.j2:
```diff
@@ -22,12 +24,20 @@ int {{ alu_name }}({{arg_list|join(',')}}, int opcode_hole_local, int immediate_
-    return x >= y;
+    return (alu_mode == 0 ? x >= y
+                          : (alu_mode == 1 ? x >= immediate_operand
+                                           : immediate_operand >= y));
```
same reason here.

taegyunkim commented:
This is also ready for review.

anirudhSK committed:
Rename file into something more appropriate.

anirudhSK committed:
XORs aren't accepted by the Tofino compiler. Switch to AND for now.

anirudhSK committed:
Fix bug in how tofino.alu's output alu is handled.

The salu ppt says that the output is either the updated value of
the register (alu_lo, alu_hi) or the original value (original_lo,
original_hi).

Note that original_lo and original_hi are called register_lo and
register_hi in the P4 code, but we use this for the updated value
of the registers because we need to update state in place in the .alu
files. Will fix this.

anirudhSK committed:
Bug fix: needs to be signed; default in P4-14 is unsigned.

Unsigned leads to unexpected behavior:

https://wesmckinney.com/blog/avoid-unsigned-integers/
https://google.github.io/styleguide/cppguide.html#Integer_Types

anirudhSK committed:
Add helper script to automate building and running for Tofino.

anirudhSK committed:
State variables should be signed as well.

anirudhSK committed:
Fix failing test.

anirudhSK committed:
Ask user for test inputs to run.sh

anirudhSK committed:
Simplify script; run should be decoupled from compile.

anirudhSK committed:
Use md5hash to avoid long table names if sketch name is long.

bfshell doesn't seem to handle long table names well and seg
faults when executing control plane commands. We need to dig deeper into
this, but for now shortening the table names seems like a reasonable
workaround.

anirudhSK committed:
The sketch_name should start with char rather than number

anirudhSK committed:
Add fix for 8b1091f1deab85f3522476714c754f3c237b76ac

taegyunkim commented on chipc/sketch_stateful_alu_visitor.py:
```diff
@@ -465,13 +465,13 @@ def generateBoolOp(self):
     return ~op2;
   }} else if (opcode == 6) {{
-    return op1 ^ op2;
+    return op1 && op2;
```
Please make a comment why we're doing this and what we really want is to model xor. 
Also update https://github.com/chipmunk-project/chipmunk-tofino/blob/b2b0ae1d779209e29273c1a2bfa17d17720e9f65/chipc/tofino_stateful_alu_visitor.py#L454

For opcode 9, too.

anirudhSK replied:
Added below.

taegyunkim committed:
Format files.

anirudhSK committed:
Add comment about XOR operations.

anirudhSK committed:
Add test for last commit.

XiangyuG commented on tests/test_tofino.py:
```diff
@@ -71,6 +71,23 @@ def test_snap_heavy_hitter_2_2_tofino_alu_cex_mode(self):
                 '1', '1', '0,1,2,3', '10']),
         )
 
+    def test_blue_increase_3_2_pred_raw_cex_mode_s0(self):
```
change the function name to test_blue_increase_3_2_tofino_alu_cex_mode_s0

XiangyuG commented on tests/test_tofino.py:
```diff
@@ -71,6 +71,23 @@ def test_snap_heavy_hitter_2_2_tofino_alu_cex_mode(self):
+            self.assertEqual(True, 'original_lo' not in content)
+            self.assertEqual(True, 'original_hi' not in content)
+
 
```
Add one more powerful test case:
```Python3
def test_blue_increase_2_2_tofino_alu_cex_mode_s0(self):
       self.assertEqual(
            0,
            iterative_solver.main([
                'iterative_solver',
                path.join(SPEC_DIR, 'blue_increase.sk'),
                path.join(STATEFUL_ALU_DIR, 'tofino.alu'),
                path.join(STATELESS_ALU_DIR, 'stateless_alu_for_tofino.alu'),
                '2', '2', '0,1,2,3', '10',
                '--state-groups', '1',
                '--target-tofino']),
        )
       with open('blue_increase_tofino_stateless_alu_for_tofino_2_2.p4') as f:
            content = f.read()
            self.assertEqual(True, 'original_lo' not in content)
            self.assertEqual(True, 'original_hi' not in content)
```

taegyunkim replied:
For your information: how to write code in github markdown: https://help.github.com/en/github/writing-on-github/basic-writing-and-formatting-syntax#quoting-code

Why is it more powerful? or rather what do you mean by powerful?

XiangyuG replied:
Oh I see. This format is much better! Thanks.

So more powerful means because the generated p4 file is different across machines and the p4 file generated from the test case test_blue_increase_3_2_tofino_alu_cex_mode_s0() did not generate the variable name **original_hi** in my machine. But this added test function will generate it. I think it will be better if we can add one more test case.

XiangyuG committed:
Add one more test case

XiangyuG committed:
Add one more space in test_tofino file
